src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Inject, Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {DOCUMENT} from '@angular/platform-browser';
+import {OverlayRef} from '../overlay-ref';
+import {Subscription} from 'rxjs/Subscription';
+import {RxChain, filter} from '@angular/cdk/rxjs';
+import {fromEvent} from 'rxjs/observable/fromEvent';
+
+/**
+ * Service for dispatching keyboard events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable()
+export class OverlayKeyboardDispatcher implements OnDestroy {
+
+  /** Currently attached overlays in the order they were attached. */
+  _attachedOverlays: OverlayRef[] = [];
+
+  private _keydownEventSubscription: Subscription | null;
+
+  constructor(@Optional() @Inject(DOCUMENT) private _document: any) { }
+
+  ngOnDestroy() {
+    if (this._keydownEventSubscription) {
+      this._keydownEventSubscription.unsubscribe();
+      this._keydownEventSubscription = null;
+    }
+  }
+
+  /** Add a new overlay to the list of attached overlay refs. */
+  add(overlayRef: OverlayRef): void {
+    // Lazily start dispatcher once first overlay is added
+    if (!this._keydownEventSubscription) {
+      this._dispatchKeydownEvents();
+    }
+
+    this._attachedOverlays.push(overlayRef);
+  }
+
+  /** Remove an overlay from the list of attached overlay refs. */
+  remove(overlayRef: OverlayRef): void {
+    const index = this._attachedOverlays.indexOf(overlayRef);
+    if (index > -1) {
+      this._attachedOverlays.splice(index, 1);
+    }
+  }
+
+  /**
+   * Subscribe to keydown events that land on the body and dispatch those
+   * events to the appropriate overlay.
+   */
+  private _dispatchKeydownEvents(): void {
```

Comment:
Judging by the name of the method, I'd assume that it would dispatch the events immediately, but that's not what it does. Something along the lines of `_subscribeToKeydownEvents` might be more appropriate.